### ISSUE TITLE
Shard server doesn't have old block when synced by root block

### ISSUE THREAD

**Issue by Bambarello** [QuarkChain/pyquarkchain#446]
Cluster fails to sync with the same error and stuck on block 24017

SLAVE_S1: E0331 07:01:37.765208 shard.py:671] Traceback (most recent call last):
SLAVE_S1: File "/code/pyquarkchain/quarkchain/cluster/shard.py", line 667, in add_block_list_for_sync
SLAVE_S1: block, skip_if_too_old=False
SLAVE_S1: File "/code/pyquarkchain/quarkchain/cluster/shard_state.py", line 842, in add_block
SLAVE_S1: block, gas_limit=gas_limit, xshard_gas_limit=xshard_gas_limit
SLAVE_S1: File "/code/pyquarkchain/quarkchain/cluster/shard_state.py", line 676, in __validate_block
SLAVE_S1: prev_header, prev_confirmed_minor_block
SLAVE_S1: File "/code/pyquarkchain/quarkchain/cluster/shard_state.py", line 549, in __is_same_minor_chain
SLAVE_S1: header.hash_prev_minor_block
SLAVE_S1: AttributeError: 'NoneType' object has no attribute 'hash_prev_minor_block'
SLAVE_S1:
SLAVE_S1: E0331 07:01:37.765900 slave.py:470] Traceback (most recent call last):
SLAVE_S1: File "slave.py", line 453, in handle_sync_minor_block_list_request
SLAVE_S1: "Failed to add minor blocks for syncing root block"
SLAVE_S1: RuntimeError: Failed to add minor blocks for syncing root block
SLAVE_S1:
MASTER: E0331 07:01:37.769117 master.py:105] Traceback (most recent call last):
MASTER: File "master.py", line 103, in sync
MASTER: await self.__run_sync()
MASTER: File "master.py", line 169, in __run_sync
MASTER: await self.__add_block(block)
MASTER: File "master.py", line 227, in __add_block
MASTER: await self.__sync_minor_blocks(root_block.minor_block_header_list)
MASTER: File "master.py", line 266, in __sync_minor_blocks
MASTER: raise RuntimeError("Unable to download minor blocks from root block")
MASTER: RuntimeError: Unable to download minor blocks from root block
MASTER:

**Comment by qizhou** [comment]
which version are you using?  In addition, it happens when you restarts your cluster or normal sync from scratch?

**Comment by qizhou** [comment]
#414 A fix should be ready here

**Comment by Bambarello** [comment]
Last version 2.4.4. The error happens when cluster restarts. When from scratch - sync is stopping at block 600

**Comment by qizhou** [comment]
Fix #414 is in the latest 2.5 testnet.  You should use that version to test as the issue is known for previous versions of testnets.

**Comment by qizhou** [comment]
Will close it as dup.  Please re-open it if you find the same issue with the latest code.  Thanks.

**Comment by ninjaahhh** [comment]
re-open the issue since the error happens for minor chain

**Comment by ninjaahhh** [comment]
confirmed the problem. it's the same symptom as #414, which will happen if a root block contains old shard blocks while the corresponding shard server doesn't have the block in memory (which is constructed during state recovery).

will put a systematic fix for this problem, both in shard server and root server (aka master server)

@Bambarello for now the temporary fix is following

```
diff --git a/quarkchain/cluster/shard_state.py b/quarkchain/cluster/shard_state.py
index 6c955af..def01da 100644
--- a/quarkchain/cluster/shard_state.py
+++ b/quarkchain/cluster/shard_state.py
@@ -546,7 +546,7 @@ class ShardState:
         header = longer_block_header
         for i in range(longer_block_header.height - shorter_block_header.height):
             header = self.db.get_minor_block_header_by_hash(
-                header.hash_prev_minor_block
+                header.hash_prev_minor_block, consistency_check=False
             )
         return header == shorter_block_header
```

**Comment by Bambarello** [comment]
> Will close it as dup. Please re-open it if you find the same issue with the latest code. Thanks.

Sorry, my typo - I'm of course running 2.5.4 version :)

**Comment by ninjaahhh** [comment]
closed by #457
